File: Final/R/weighted_least_squares.R
library(car)
library(GGally)
library(ggplot2)

df <- read.csv('../Data/combined.csv')

# +/- is rendered as X...
# E+/- is rendered as E...
model_R1_log <- lm(log(Salary) ~ GP + PTS + X... + Grit + PIM, data = df)
summary(model_R1_log)
autoplot(model_R1_log)

e_i_R1    <- resid(model_R1_log)
y_hat_R1  <- predict(model_R1_log)
k_hat     <- summary(model_R1_log)$sigma^2

plot(y_hat_R1, abs(rstudent(model_R1_log)))
plot(df$GP, abs(rstudent(model_R1_log)))
plot(df$PTS, abs(rstudent(model_R1_log)))
plot(df$X..., abs(rstudent(model_R1_log)))
plot(df$Grit, abs(rstudent(model_R1_log)))
plot(df$PIM, abs(rstudent(model_R1_log)))

# Try with Games Played (GP)
##### standard deviation function #####
sd_fun_data   <- data.frame(y = abs(e_i_R1), x = df$GP)
model_R1_sdf  <- lm(y ~ x, data = sd_fun_data)
sigma_hat     <- predict(model_R1_sdf)
w_i_sd        <- 1/(sigma_hat^2)

model_R1_sd_wls <- lm(log(Salary) ~ GP + PTS + X... + Grit  + PIM, 
                      data = df, weights = w_i_sd)
round(coef(model_R1_log), 5)
round(coef(model_R1_sd_wls), 5)
autoplot(model_R1_sd_wls)
mean(vif(model_R1_sd_wls))
summary(model_R1_sd_wls)

File: Final/weighted_least_squares.py
# %%
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import OLSInfluence

from src.nhl_linreg import nhl_linreg
# %%
df = pd.read_csv('Data/combined.csv', encoding='latin-1')

# %%
Y = df['Salary']
cols = ['GP', 'PTS', '+/-', 'Grit', 'TOI', 'PIM', 'Is_Offense']
X = df[cols]
X = sm.add_constant(X)

model3 = sm.OLS(Y, X)
results3 = model3.fit()
print(results3.summary())
print(results3.rsquared_adj)
diag3 = nhl_linreg(results3)
vif3, fig3, ax3 = diag3()
print(vif3['VIF Factor'].mean()) # 7.22

# %%
e_i = results3.resid
y_hat = results3.fittedvalues
k_hat = results3.scale

influence = OLSInfluence(results3)
rstudent = influence.get_resid_studentized_external()

# %%
fig, ax = plt.subplots()
ax.scatter(y_hat, np.abs(rstudent))
ax.set_title('Y-hat vs. |studentized residuals|')
plt.show();

# %%
fig, ax = plt.subplots()
ax.scatter(df['GP'], np.abs(rstudent))
ax.set_title('Games Played vs. |studentized residuals|')
plt.show();

# %%
fig, ax = plt.subplots()
ax.scatter(df['PTS'], np.abs(rstudent))
ax.set_title('Points vs. |studentized residuals|')
plt.show();

# %%
fig, ax = plt.subplots()
ax.scatter(df['+/-'], np.abs(rstudent))
ax.set_title('+/- vs. |studentized residuals|')
plt.show();

# %%
fig, ax = plt.subplots()
ax.scatter(df['Grit'], np.abs(rstudent))
ax.set_title('Grit vs. |studentized residuals|')
plt.show();

# %%
fig, ax = plt.subplots()
ax.scatter(df['TOI'], np.abs(rstudent))
ax.set_title('Time on ice vs. |studentized residuals|')
plt.show();

# %%
fig, ax = plt.subplots()
ax.scatter(df['PIM'], np.abs(rstudent))
plt.show();

# %%
fig, ax = plt.subplots()
ax.scatter(df['Is_Offense'], np.abs(rstudent))
plt.show();

# %%
# trying to improve GP
# standard deviation function
sd_fun_data = pd.DataFrame({
    'y': np.abs(e_i),
    'x': df['GP']
})
Y = sd_fun_data['y']
X = sd_fun_data['x']
model_sdf = sm.OLS(Y, X)
results_sdf = model_sdf.fit()
sigma_hat = results_sdf.fittedvalues
w_i_sd  = 1/(sigma_hat ** 2)

# %%
Y = df['Salary']
cols = ['GP', 'PTS', '+/-', 'Grit', 'TOI', 'PIM', 'Is_Offense']
X = df[cols]
X = sm.add_constant(X)

model_wls = sm.WLS(Y, X, weights = w_i_sd)
results_wls = model_wls.fit()
print(results_wls.summary())